src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -50,69 +54,127 @@
 
     protected abstract List<FieldSignature> staticFieldSignatures();
 
-    // TODO should precompute hash sets of signatures, assuming we document that the signatures may not change over the lifetime of the whitelist (or pass them in the constructor)
+    ConcurrentHashMap<String, Boolean> permittedCache = new ConcurrentHashMap<String, Boolean>();  // Not private to facilitate testing
+
+    private void cacheSignatureList(List<Signature> ...sigs) {
+        for (List<Signature> list : sigs) {
+            for (Signature s : list) {
+                if (!s.isWildcard()) { // Cache entries for wildcard signatures will never be accessed and just waste space
+                    permittedCache.put(s.toString(), Boolean.TRUE);
+                }
+            }
+        }
+    }
+
+    /** Prepopulates the "permitted" cache. */
+    public void precache() {
+        cacheSignatureList((List)methodSignatures(), (List)(newSignatures()),
+                           (List)(staticMethodSignatures()), (List)(fieldSignatures()),
+                           (List)(staticFieldSignatures()));
+    }
+
+    /** Frees up memory used for the cache. */
+    void clearCache() {
+        this.permittedCache.clear();
+        this.permittedCache = new ConcurrentHashMap<String, Boolean>();
```

Comment:
What is the point of having _both_ these statements?

Reply:
@jglick HashMaps don't shrink once expanded and there's no `arrayList.trimToSize()` equivalent.